### PR TITLE
Faster diffs: Don't always run pygit2.Tree.diff_tree

### PR DESCRIPTION
**Pull request by olsen232**
Don't run it when the user requests the diff of a particular feature
- we can look up the feature for both revs quicker than pygit2 can find all diffs in the entire revision.

There will be an extremely great speedup in the extreme case where the user wants to see a single feature that has changed, across a revision-range where a million features have changed.

If the user requests enough features, eventually there will be a crossover where this code path is slower. Eg if the user requests to see more features than have actually changed. However, mostly the size of the list of features requested will be much smaller than the size of the layer / size of the changes.

## Checklist:

- [x] Have you reviewed your own change?
- [x] Have you included test(s)?
- [x] Have you updated the [changelog](https://github.com/koordinates/kart/blob/master/CHANGELOG.md)?

### CODE REVIEW

File: kart/dataset_mixins.py [hunk 1]
```diff
@@ -160,6 +166,66 @@ def get_decoder(dataset, method_name):
     _INSERT_UPDATE = _INSERT_TYPES + _UPDATE_TYPES
     _UPDATE_DELETE = _UPDATE_TYPES + _DELETE_TYPES
 
+    def get_raw_deltas_for_keys(
+        self, other, key_encoder_method, key_filter, reverse=False
```
Comment (craigds) [inline kart/dataset_mixins.py:170]: type hints on args wouldn't go amiss, the names don't make them very obvious

(we should probably be moving towards type hints everywhere)

File: tests/test_diff.py [hunk 2]
```diff
@@ -1581,7 +1582,17 @@ def test_show_points_HEAD(output_format, data_archive_readonly, cli_runner):
         ["show", "HEAD", "--", "nz_pa_points_topo_150k:1182"],
     ],
 )
-def test_diff_filtered_text(diff_command, data_archive_readonly, cli_runner):
+def test_diff_filtered_text(
+    diff_command, data_archive_readonly, cli_runner, monkeypatch
+):
+    def _get_raw_diff_for_subtree(self, *args, **kwargs):
+        # When only nz_pa_points_topo_150k:1182 is requested, a different more efficient code-path should be used.
+        raise AssertionError(
```
Comment (craigds) [inline tests/test_diff.py:1590]:
```suggestion
        pytest.fail(
```

**Review by craigds — APPROVED** [review]

File: kart/dataset_mixins.py (hunk 1 above)
Comment (olsen232) [reply kart/dataset_mixins.py]: Done - added type hints to this file

File: tests/test_diff.py (hunk 2 above)
Comment (olsen232) [reply tests/test_diff.py]: Done

**Pull request merged**